pkg/migration-tool/utils/client.go

```diff
@@ -119,26 +195,47 @@ func (c *Client) Read(ctx context.Context, query *prompb.Query, desc string) (re
 	}
 	compressed, err = ioutil.ReadAll(bytes.NewReader(reader.Bytes()))
 	if err != nil {
-		return nil, -1, -1, errors.Wrap(err, fmt.Sprintf("error reading response. HTTP status code: %s", httpResp.Status))
+		err = errors.Wrap(err, fmt.Sprintf("error reading response. HTTP status code: %s", httpResp.Status))
+		if shouldRetry(c.cRuntime.OnErr, c.cRuntime, numAttempts, fmt.Sprintf("error:\n%s\n", err.Error())) {
+			goto retry
+		}
+		return nil, -1, -1, err
 	}
 
 	if httpResp.StatusCode/100 != 2 {
+		if shouldRetry(c.cRuntime.OnErr, c.cRuntime, numAttempts, fmt.Sprintf("received status code: %d", httpResp.StatusCode)) {
+			goto retry
+		}
 		return nil, -1, -1, errors.Errorf("remote server %s returned HTTP status %s: %s", c.url.String(), httpResp.Status, strings.TrimSpace(string(compressed)))
 	}
 
 	uncompressed, err := snappy.Decode(nil, compressed)
 	if err != nil {
+		// Snappy errors also happen in case of deadline or context cancel since the read was terminated in middle of the process.
```

Comment:
No, we don't. I too expected the same behaviour, but in testing, it didn't send the error above rather as corrupt snappy since the timeout occurred. The snappy corruption is because prom-migrator accepts response, and while accepting timeout occurs and gets a partial snappy encoded response, which is bad.

Reply:
Yes, when I tried with local Prometheus, `if errors.Is(err, context.DeadlineExceeded) {` is catching the deadline issue and then retrying. So, we can get rid of this exception at line 214 which you commented. Thank you.